### PR TITLE
apis: add UpstreamTraffic API types

### PR DESCRIPTION
**Pull request by shashankram**
**Description**:
Adds the Go types for  UpstreamTraffic API.

*Note the defaults are not to be encoded in
the JSON tags and will be implemented within
the controller subsequently.

Part of #4500

**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| New Functionality          | [X] |

Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project? `no`
    -   Did you notify the maintainers and provide attribution?

2. Is this a breaking change? `no`

### CODE REVIEW

File: pkg/apis/policy/v1alpha1/upstreamtraffic.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	// +optional
+	HTTP *HTTPConnectionSettings `json:"http,omitempty"`
+}
+
+// TCPConnectionSettings defines the TCP connection settings for an
+// upstream host.
+type TCPConnectionSettings struct {
+	// MaxConnections specifies the maximum number of TCP connections
+	// allowed to the upstream host.
+	// Defaults to 1024 if not specified.
```
Comment (nojnhuh) [inline pkg/apis/policy/v1alpha1/upstreamtraffic.go:58]: I wonder if it's worth distinguishing whether it's OSM or Envoy who's actually applying this default. I could interpret this as "if I don't specify this when the resource is created and I read it later, it will have the default" (where OSM is expected to apply the default), but if we simply pass the value to Envoy and let it apply the default then that wouldn't be the case.

File: pkg/apis/policy/v1alpha1/upstreamtraffic.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	// +optional
+	MaxConnections uint32 `json:"maxConnections,omitempty"`
```
[the remark this answers is not in the document]
Comment (shashankram) [reply pkg/apis/policy/v1alpha1/upstreamtraffic.go:60]: The fields with finite defaults do not accept values that mean unlimited. Most of these have a ceil of 2^32 - 1.

The behavior of defaults is completely left to OSM regardless of Envoy defaults. 0 is not equivalent to not specified unless stated so. A validating webhook will prevent such ambiguity from occurring.

**Review by snehachhabria — APPROVED** [review]

File: pkg/apis/policy/v1alpha1/upstreamtraffic.go (hunk 1 above)
Comment (shashankram) [reply pkg/apis/policy/v1alpha1/upstreamtraffic.go:58]: This is an OSM API, not Envoy's. So the defaults will be applied by OSM controller regardless of what Envoy uses as defaults. This is primarily because some Envoy defaults may not be suitable defaults. At present, I have set these to match Envoy defaults, but these could change later, e.g. to 2^32 - 1 based on the most common use case. At the moment it's hard to reason how sensible these defaults are. Given the fields are optional in Envoy with defaults values if unspecified, I prefer to explicitly specify defaults from within OSM for such fields.

File: pkg/apis/policy/v1alpha1/upstreamtraffic.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (nojnhuh) [reply pkg/apis/policy/v1alpha1/upstreamtraffic.go:60]: Is zero still a valid value? If so, doesn't this field need to be a pointer then so we can differentiate between specified (use the given value) and unspecified (use the default)?

[the remark this answers is not in the document]
Comment (shashankram) [reply pkg/apis/policy/v1alpha1/upstreamtraffic.go:60]: Zero is not a valid value unless specified. A validating webhook will prevent invalid values. The controller will only override the default if this is non-zero.

**Review by jaellio — APPROVED** [review]

File: pkg/apis/policy/v1alpha1/upstreamtraffic.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (nojnhuh) [reply pkg/apis/policy/v1alpha1/upstreamtraffic.go:60]: > A validating webhook will prevent invalid values.

If all the webhook sees is a `uint32` that equals 0, how does it know whether or not that was specified? That's why I think this needs to be a pointer where `nil` means unspecified.

File: pkg/apis/policy/v1alpha1/upstreamtraffic.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+	// requests allowed to the upstream host.
+	// Defaults to 1024 if not specified.
+	// +optional
+	MaxPendingRequests uint32 `json:"maxPendingRequests,omitempty"`
+
+	// MaxRetries specifies the maximum number of parallel retries
+	// allowed to the upstream host.
+	// Defaults to 3 if not specified.
+	// +optional
+	MaxRetries uint32 `json:"maxRetries,omitempty"`
```
Comment (steeling) [inline pkg/apis/policy/v1alpha1/upstreamtraffic.go:93]: To Jon's point, wouldn't 0 be a valid value for MaxRetries? If that's the case then the statement "Defaults to 3 if not specified" is incorrect, since we can't distinguish between 0 and not specified

File: pkg/apis/policy/v1alpha1/upstreamtraffic.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// UpstreamTraffic defines the policy applicable to traffic destined
+// to an upstream host.
+// +genclient
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type UpstreamTraffic struct {
```
Comment (steeling) [inline pkg/apis/policy/v1alpha1/upstreamtraffic.go:12]: COmmented on the design, but commenting here for posterity. Traffic would indicate the actual bits flowing through the system, whereas this is Setting for upstream traffic. Therefore I think a more appropriate name would be UpstreamTrafficSetting

File: pkg/apis/policy/v1alpha1/upstreamtraffic.go (hunk 3 above)
Comment (shashankram) [reply pkg/apis/policy/v1alpha1/upstreamtraffic.go:93]: At this point, there is no implementation backing the API to render defaults. The controller would have logic like
`if (maxRetries > 0) // configure max retries`. I don't yet know whether this can be set to zero, needs testing, so at this point made the assumption that a non-zero value will not be allowed for max_retries similar to the other uint32 values.

File: pkg/apis/policy/v1alpha1/upstreamtraffic.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (shashankram) [reply pkg/apis/policy/v1alpha1/upstreamtraffic.go:60]: The CRD schema will enforce a min value of 1.

File: pkg/apis/policy/v1alpha1/upstreamtraffic.go (hunk 4 above)
Comment (shashankram) [reply pkg/apis/policy/v1alpha1/upstreamtraffic.go:12]: I chose to keep this name concise similar to other API Kinds, e.g [Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/), [Egress](https://release-v1-0.docs.openservicemesh.io/docs/api_reference/policy/v1alpha1/#policy.openservicemesh.io/v1alpha1.Egress) etc. I can see how renaming this to `UpstreamTrafficSetting` would make this less ambiguous, so I'll rename it.

File: pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// UpstreamTrafficSetting defines the settings applicable to traffic destined
+// to an upstream host.
```
Comment (draychev) [inline pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go:8]: Examples being Circuit Breaking

**Review by draychev — APPROVED** [review]
This is good stuff! Glad to see this being built out!
Thanks @shashankram !
